OpenSim/Simulation/StatesTrajectory.h

```diff
@@ -0,0 +1,641 @@
+     * the in-memory time of a state is actually 1.500001 but it is printed to
+     * the console as 1.500; you might be tempted to get this state by asking
+     * for the one that is before/at time 1.5, but this would end up giving you
+     * the state at time 1.3. You could remedy this issue by setting the
+     * tolerance to 1e-3 (the precision of the console output).
+     *
+     * @throws TimeOutOfRange If there are no states before the given time.
+     * @throws Exception If the trajectory is empty.
+     */
+    size_t getIndexBefore(const double& time, const double& tolerance);
+    /** Same as getIndexAfter(), but accepts a tolerance to handle imprecise
+     * times.
+     * See getIndexBefore(const double&, const double&) for an explanation.
+     *
+     * @throws TimeOutOfRange If there are no states after the given time.
+     * @throws Exception If the trajectory is empty.
+     */
+    size_t getIndexAfter(const double& time, const double& tolerance);
+    /** Same as getBetween(), but accepts a tolerance to handle
+     * imprecise times.
+     * See getIndexBefore(const double&, const double&) for an explanation.
+     */
+    IteratorRange getBetween(const double& startTime, const double& endTime, 
+            const double& tolerance);
+    /** Same as getIteratorBefore(), but accepts a tolerance to handle
+     * imprecise times.
+     * See getIndexBefore(const double&, const double&) for an explanation.
+     */
+    const_iterator getIteratorBefore(const double& time,
+            const double& tolerance);
+    /** Same as getIteratorAfter(), but accepts a tolerance to handle
+     * imprecise times.
+     * See getIndexBefore(const double&, const double&) for an explanation.
+     */
+    const_iterator getIteratorAfter(const double& time,
+            const double& tolerance);
+    /// @}
+
+private:
+
+    /** Checks if two states have the same number of state variables,
+     * constraints, etc. Returns true if the following quantities are the same
+     * for both states:
+     * - number of generalized coordinates (Q's)
+     * - number of generalized speeds (U's)
+     * - number of auxiliary state variables (Z's)
+     * - number of position constraints (QErr's)
+     * - number of velocity constraints (UErr's)
+     * - number of acceleration constraints (UDotErr's)
+     * - number of constraint Lagrange multipliers
+     * - number of event triggers
+     *
+     * Returns false otherwise.
+     *
+     * This method does NOT check for any relationship between the times in the
+     * two states.
+     */
+    // TODO this function should be pushed to the SimTK::State class, so that
+    // the check can evolve with the State class.
+    static bool isConsistent(const SimTK::State& stateA,
+                             const SimTK::State& stateB);
+
+    std::vector<SimTK::State> m_states;
+
+public:
+
+    /** Thrown when asking for a state at a time that is out of range. */
```

Comment:
Nested classes are shown at the top in Doxygen and it would be great if it had a heading like: "Exceptions thrown by StatesTrajectory" or "StatesTrajectory Exceptions". I really like these!